libflux/src/core/ast/mod.rs

```diff
@@ -492,8 +492,131 @@ pub struct BuiltinStmt {
+pub enum MonoType {
+    Tvar(TvarType),
+    Int(Int),
+    Uint(Uint),
+    Float(Float),
+    String(StringType),
+    Bool(Bool),
+    Time(Time),
+    Duration(DurationType),
+    Bytes(Bytes),
+    Regexp(Regexp),
+    Invalid,
+}
+
+pub fn base_from_monotype(m : &MonoType) -> BaseNode{
+    match m {
+        MonoType::Tvar(t) => t.base.clone(),
+        MonoType::Int(t) => t.base.clone(),
+        MonoType::Uint(t) => t.base.clone(),
+        MonoType::Float(t) => t.base.clone(),
+        MonoType::String(t) => t.base.clone(),
+        MonoType::Bool(t) => t.base.clone(),
+        MonoType::Time(t) => t.base.clone(),
+        MonoType::Duration(t) => t.base.clone(),
+        MonoType::Bytes(t) => t.base.clone(),
+        MonoType::Regexp(t) => t.base.clone(),
+        _ =>  BaseNode::default(),
+    }
+}
+
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct TvarType {
+    #[serde(skip_serializing_if = "BaseNode::is_empty")]
+    #[serde(default)]
+    #[serde(flatten)]
+    pub base: BaseNode,
+}
+
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct Int {
+    #[serde(skip_serializing_if = "BaseNode::is_empty")]
+    #[serde(default)]
+    #[serde(flatten)]
+    pub base: BaseNode,
+}
+
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct Uint {
+    #[serde(skip_serializing_if = "BaseNode::is_empty")]
+    #[serde(default)]
+    #[serde(flatten)]
+    pub base: BaseNode,
+}
+
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct Float {
+    #[serde(skip_serializing_if = "BaseNode::is_empty")]
+    #[serde(default)]
+    #[serde(flatten)]
+    pub base: BaseNode,
+}
+
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct StringType {
+    #[serde(skip_serializing_if = "BaseNode::is_empty")]
+    #[serde(default)]
+    #[serde(flatten)]
+    pub base: BaseNode,
+}
+
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct Bool {
+    #[serde(skip_serializing_if = "BaseNode::is_empty")]
+    #[serde(default)]
+    #[serde(flatten)]
+    pub base: BaseNode,
+}
+
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct Time {
+    #[serde(skip_serializing_if = "BaseNode::is_empty")]
+    #[serde(default)]
+    #[serde(flatten)]
+    pub base: BaseNode,
+}
+
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct DurationType {
+    #[serde(skip_serializing_if = "BaseNode::is_empty")]
+    #[serde(default)]
+    #[serde(flatten)]
+    pub base: BaseNode,
+}
+
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct Bytes {
+    #[serde(skip_serializing_if = "BaseNode::is_empty")]
+    #[serde(default)]
+    #[serde(flatten)]
+    pub base: BaseNode,
+}
+
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct Regexp {
+    #[serde(skip_serializing_if = "BaseNode::is_empty")]
+    #[serde(default)]
+    #[serde(flatten)]
+    pub base: BaseNode,
+}
```

Comment:
Since each of these structs contains only a BaseNode field would it make sense to create a single `BasicType` node like this?

```
#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub struct BasicType {
    #[serde(skip_serializing_if = "BaseNode::is_empty")]
    #[serde(default)]
    #[serde(flatten)]
    pub base: BaseNode,
    pub name: Identifier,
}
```

Then the specific basic type is just stored as an identifier on the node? 

There are pros and cons to either pattern curious what others think?

Reply:
@jlapacik Thoughts :point_up: ?

Reply:
I think that removes a lot of redundancy. Are you thinking something like this?
```Rust
pub enum MonoType {
    Basic(Basic),
    ...
}

pub struct Basic {
    pub base: BaseNode,
    pub ty: Primitive,
}

pub enum Primitive {
    Int,
    Uint,
    ...
}
```

Reply:
I think keeping them as Identifiers is best because we don't have any vanilla enum values like the Primitive enum in the AST and it creates some challenges with serialization. For example what would the JSON be for the Primitive enum? Is it a number or a string? A numer would make working with the AST in other languages difficult because you would need know the mapping of number to enum. If its a string thats basically the same thing as an Identifier so I say just use that.

Reply:
@yuktarakesh @ChitlangeSahas I think @nathanielc's suggestion makes for a good simplification. Especially when it comes time to implement the serialization bits.

Reply:
One other thought I had here, should we call the AST node NamedType instead of BasicType? That way if we ever have user defined types we can use the same ast node to represent it since its just a type named by its identifier?